internal/moreos/test/moreos_linux_test.go

```diff
@@ -0,0 +1,105 @@
+)
+
+// TestProcessGroupAttr_Kill sends SIGKILL to a running fake func-e that spawns a fake envoy process.
+func TestProcessGroupAttr_Kill(t *testing.T) {
+	tempDir := t.TempDir()
+	fakeFuncE := filepath.Join(tempDir, "func-e")
+	fakeEnvoyDir := filepath.Join(tempDir, "versions", string(version.LastKnownEnvoy))
+	fakeEnvoy := filepath.Join(fakeEnvoyDir, "envoy")
+	require.NoError(t, os.MkdirAll(fakeEnvoyDir, 0755))
+
+	test.RequireFakeFuncE(t, fakeFuncE+moreos.Exe)
+	test.RequireFakeEnvoy(t, fakeEnvoy+moreos.Exe)
+
+	cmd := exec.CommandContext(context.Background(), fakeFuncE)
+	cmd.Env = []string{"FUNC_E_HOME=" + tempDir}
+	stderr := new(bytes.Buffer)
+	cmd.Stderr = stderr
+	require.NoError(t, cmd.Start())
+
+	// Block until we reach an expected line or timeout.
+	reader := bufio.NewReader(stderr)
+	waitFor := "initializing epoch 0"
+	if !assert.Eventually(t, func() bool {
+		b, e := reader.Peek(512)
+		return e != nil && strings.Contains(string(b), waitFor)
+	}, 5*time.Second, 100*time.Millisecond) {
+		require.FailNowf(t, "timeout waiting for stderr to contain %q", waitFor)
+	}
+
+	fakeFuncEProcess, err := process.NewProcess(int32(cmd.Process.Pid))
+	require.NoError(t, err)
+
+	// Get all fake func-e children processes.
+	children, err := fakeFuncEProcess.Children()
+	require.NoError(t, err)
+	require.Equal(t, len(children), 1) // Should have only one child process i.e. the fake envoy process.
+	fakeEnvoyProcess := &os.Process{Pid: int(children[0].Pid)}
+	requireFindProcessNoError(t, fakeEnvoyProcess)
+
+	// Kill the fake func-e process.
+	// This works only for linux, sending kill -9 on darwin will not kill the process, we need to kill
+	// via pgid or kill the child first.
+	require.NoError(t, cmd.Process.Kill())
+	// Wait for the process to die; this could error due to the kill signal.
+	cmd.Wait() //nolint
+
+	// Wait and check if fake func-e and envoy processes are killed.
+	requireFindProcessError(t, cmd.Process, process.ErrorProcessNotRunning)
+	requireFindProcessError(t, fakeEnvoyProcess, process.ErrorProcessNotRunning)
+
+	// Ensure both processes are killed.
+	require.NoError(t, moreos.EnsureProcessDone(cmd.Process))
+	require.NoError(t, moreos.EnsureProcessDone(fakeEnvoyProcess))
+}
+
+func requireFindProcessError(t *testing.T, proc *os.Process, expectedErr error) {
+	// Wait until the operating system removes or adds the scheduled process.
+	if !assert.Eventually(t, func() bool {
+		_, err := process.NewProcess(int32(proc.Pid)) // because os.FindProcess is no-op in Linux!
+		return err == expectedErr
+	}, 100*time.Millisecond, 5*time.Millisecond) {
+		if expectedErr == nil {
+			require.FailNow(t, "timeout waiting for finding process with no error")
+		}
+		require.FailNow(t, "timeout waiting for expected error %v", expectedErr)
+	}
+}
+
+func requireFindProcessNoError(t *testing.T, proc *os.Process) {
+	requireFindProcessError(t, proc, nil) // expect to find the process with no error.
+}
```

Comment:
This part while there will be some kinks is pretty on the money. I will have some comments to make, but nothing huge.

Reply:
What do you mean by: "This part while there will be some kinks is pretty on the money."? My English seems bad to capture the real meaning of it. But if you have comments, please add them. I'll be happy to learn from it.

Reply:
sorry by "on the money" I meant exactly right. I prefixed "pretty" because a couple nits, but I didn't want to nit while some other things are in play.

Reply:
so translating from Adrian into English "This part while there will be some kinks is pretty on the money."  "this part is almost exactly right, and we can save the minor details for later" :D